documentation/smg/16-serialiser.md

```diff
@@ -0,0 +1,275 @@
+# Serialisation details
+
+As of the integration of the C++ serialiser, the serialised format has been updated to be more consistent and to have greater capabilities.
+
+For reference, the old format is also detailed below
+
+## New format
+
+### Standard format
+The new style has a standardised header format depending on the shape of the data (with the exception of function handles).
+This format allows extension to support arrays of objects and structs in a way which the old serialiser couldn't.
+It also greatly simplifies the tag structure and allows it to align with the C++ MEX API data types.
+Data in these cases is serialised in the linearised order as defined by `data(:)` and the array reshaped by the deserialiser.
+
+This standard format for the header is as follows:
+RANK |     tag    | dim1 | dim2 | ... | Data
+-----|------------|------|------|-----|------
+NULL | 32 + tag   | 0    |      |     |
+0    |    tag     |      |      |     | Data
+1    | 32 + tag   | nElem|      |     | Data
+2    | 64 + tag   | dim1 | dim2 |     | Data
+N    | N<<5 + tag | dim1 | dim2 | ... | Data
+
+Where `<<5` means bitshift left 5. (see [Tag Format][tag])
+
+- Tags are `uint8` data and the "magic" numbers are described in detail in the section [Tag Format][tag] below.
+- Dimensions are serialised as `uint32` 
+
+The single exception to this header is that of `function_handles` which are purely scalar and the rank component of the tag is used to specify the type of function handle:
+
+Handle type         |   tag   | value
+--------------------|---------|-------
```

Comment:
So here we are looking at bit patterns (with 25=11001)
00111001
01011001
01111001
checking if this is the case. As the value is the sum of 25 and the tag bit position, presuming it is not a separately stored field - does it make sense to provide the value separately?

Reply:
That is correct. I guess we could, but not sure we need to

Reply:
Oh sorry, reread the question and it makes more sense, I was originally planning on also writing a quick reference for each, but decided against it as it is unlikely anyone who is going to be manually decoding serialised data will not be able to do a simple addition.